chain-tx-enclave/enclave-ra/ra-client/src/verifier.rs

```diff
@@ -0,0 +1,226 @@
+        let (_, certificate) = parse_x509_der(certificate)
+            .map_err(|_| EnclaveCertVerifierError::CertificateParsingError)?;
+
+        if certificate
+            .tbs_certificate
+            .validity
+            .time_to_expiration()
+            .is_none()
+        {
+            return Err(EnclaveCertVerifierError::CertificateExpired);
+        }
+
+        let attestation_report_oid = Oid::from(OID_EXTENSION_ATTESTATION_REPORT);
+        let mut attestation_report_received = false;
+
+        let public_key_hash = sha256(
+            certificate
+                .tbs_certificate
+                .subject_pki
+                .subject_public_key
+                .data,
+        );
+
+        for extension in certificate.tbs_certificate.extensions {
+            if extension.oid == attestation_report_oid {
+                attestation_report_received = true;
+                self.verify_attestation_report(extension.value, public_key_hash)?;
+            }
+        }
+
+        if attestation_report_received {
+            Ok(())
+        } else {
+            Err(EnclaveCertVerifierError::MissingAttestationReport)
+        }
+    }
+
+    /// Verifies attestation report
+    fn verify_attestation_report(
+        &self,
+        attestation_report: &[u8],
+        public_key_hash: [u8; 32],
+    ) -> Result<(), EnclaveCertVerifierError> {
+        log::info!("Verifying attestation report");
+
+        let trust_anchors: Vec<TrustAnchor> = self
+            .root_cert_store
+            .roots
+            .iter()
+            .map(|cert| cert.to_trust_anchor())
+            .collect();
+        let time =
+            Time::try_from(SystemTime::now()).map_err(|_| EnclaveCertVerifierError::TimeError)?;
+
+        let attestation_report: AttestationReport = bincode::deserialize(attestation_report)?;
+
+        let signing_certs = certs(&mut attestation_report.signing_cert.as_ref())
+            .map_err(|_| EnclaveCertVerifierError::CertificateParsingError)?;
+
+        for signing_cert in signing_certs {
+            let signing_cert = EndEntityCert::from(&signing_cert.0)?;
+
+            signing_cert.verify_is_valid_tls_server_cert(
+                SUPPORTED_SIG_ALGS,
+                &TLSServerTrustAnchors(&trust_anchors),
+                &[],
+                time,
+            )?;
+
+            signing_cert.verify_signature(
+                &RSA_PKCS1_2048_8192_SHA256,
+                &attestation_report.body,
+                &attestation_report.signature,
+            )?;
+        }
+
+        self.verify_attestation_report_body(&attestation_report.body, public_key_hash)?;
+
+        log::info!("Attestation report is valid!");
+        Ok(())
+    }
+
+    fn verify_attestation_report_body(
+        &self,
+        attestation_report_body: &[u8],
+        public_key_hash: [u8; 32],
+    ) -> Result<(), EnclaveCertVerifierError> {
+        let attestation_report_body: AttestationReportBody =
```

Comment:
As discussed with @tomtau on slack, the only thing that we can change is the usage of bincode. I'll change the code to use `serde_json` instead of `bincode` to reduce the number of dependencies and make everything backwards compatible.

Reply:
Done.